### DISCUSSION TITLE
TLS in TLS detection in: Vless Reality ==> OpenVPN ==> TOR relay?

### DISCUSSION THREAD

**Discussion by Morphey3020** [XTLS/Xray-core#3422] [Q&A] [answered]
I haven't fully figured out the detection of TLS in TLS. So I'm wondering if this scheme is vulnerable to TLS in TLS detection: VLESS XTLS-REALITY VISION ==> OpenVPN (UDP) ==> TOR relay?

**Comment by kingwilliam** [comment, marked as answer]
TLS sequence 5 packet and direction, like Morse Code
```
1. very short		(client -> server)
2. very short		(client <- server)
3. short		(client -> server)
4. very long		(client <- server)
5/ very short		(client -> server)
```
this is simple way to understand TLS IN TLS

so, OpenVPN + tor only just encrypt + encrypt packet, but cannot change 5 packet and direction


if you have interest to understand more details in TLS in TLS, you can visit

original post
[XTLS Vision, fixes TLS in TLS, to the star and beyond #1295](https://github.com/XTLS/Xray-core/discussions/1295)

translated to english
[https://github.com/seakfind/examples/blob/main/xtls-vision/README.md](https://github.com/seakfind/examples/blob/main/xtls-vision/README.md)
